src/bootchooser.c

```diff
@@ -1375,6 +1375,206 @@ static gboolean efi_get_state(RaucSlot* slot, gboolean *good, GError **error)
 	return TRUE;
 }
 
+/* Wrapper for get commands accessing custom script
+ *
+ * @param cmd What to input as command to the custom backend. Mandatory.
+   Valid values:
+   - get-primary
+   - get-state
+ * @param bootname slot.bootname
+ * @param ret_str Return string from stdout after running the command
+ * @param error Return location for a GError
+ */
+static gboolean custom_backend_get(const gchar *cmd, gchar *bootname, GString **ret_str, GError **error)
+{
+	g_autoptr(GSubprocess) sub = NULL;
+	GError *ierror = NULL;
+	g_autoptr(GBytes) stdout_buf = NULL;
+	gint ret;
+	gsize size;
+	const char *data;
+	gchar *backend_name = r_context()->config->custom_bootloader_backend;
+
+	g_return_val_if_fail(cmd, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (bootname)
+		sub = g_subprocess_new(G_SUBPROCESS_FLAGS_STDOUT_PIPE, &ierror, backend_name, cmd, bootname, NULL);
+	else
+		sub = g_subprocess_new(G_SUBPROCESS_FLAGS_STDOUT_PIPE, &ierror, backend_name, cmd, NULL);
+
+	if (!sub) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to start : %s", backend_name);
+		return FALSE;
+	}
+
+	if (!g_subprocess_communicate(sub, NULL, NULL, &stdout_buf, NULL, &ierror)) {
+		g_propagate_prefixed_error(
+				error,
+				ierror,
+				"Failed to run : %s", backend_name);
+		return FALSE;
+	}
+	data = g_bytes_get_data(stdout_buf, &size);
+	*ret_str = g_string_new(data);
+	// Cleanup string for newlines
+	g_strchomp((*ret_str)->str);
```

Comment:
Hm. You could also use g_strstrip() to also remove leading whitespace.

But how does a GString object react to having its ->str member modified without updating the ->len member? OK, the callers only seem to use the ->str member, but perhaps that means the GString container is the wrong thing to use here - maybe just take a "const char **ret_str" and create that with g_strndup() ?

Reply:
converted to use gchar instead of GString